lxd/storage/backend_lxd.go

```diff
@@ -2223,6 +2225,225 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
+
+			// Create temporary file where image is uploaded.
+			file, err := os.Create(imgUploadPath)
+			if err != nil {
+				return fmt.Errorf("Failed to create temporary file (%s): %w", imgUploadPath, err)
+			}
+
+			// Remove uploaded image regardless of image conversion success.
+			defer file.Close()
+			defer os.Remove(imgUploadPath)
+		}
+
+		args.ConversionFilePath = imgUploadPath
+
+		// Receive/upload image.
+		err = b.driver.CreateVolumeFromMigration(drivers.NewVolumeCopy(vol), conn, args, nil, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _ = b.DeleteInstance(inst, op) })
+
+		// Mount volume and get disk path.
+		err = b.driver.MountVolume(vol, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _, _ = b.driver.UnmountVolume(vol, false, op) })
+		diskPath, err := b.driver.GetVolumeDiskPath(vol)
+		if err != nil {
+			return fmt.Errorf("Failed getting instance volume disk path: %v", err)
+		}
+
+		// If imgUploadPath is empty, it means that the image was uploaded directly into the instance
+		// volume. For example, if image conversion is not requested. In such case, use diskPath as
+		// imgUploadPath.
+		if imgUploadPath == "" {
+			imgUploadPath = diskPath
+		}
+
+		// Extract image format and size.
+		cmd := []string{"qemu-img", "info", imgUploadPath, "--output", "json"}
+		out, err := apparmor.QemuImg(b.state.OS, cmd, imgUploadPath, "")
+		if err != nil {
+			return fmt.Errorf("qemu-img info: %v", err)
+		}
+
+		imgInfo := struct {
+			Format string `json:"format"`
+			Bytes  int64  `json:"virtual-size"`
+		}{}
+
+		err = json.Unmarshal([]byte(out), &imgInfo)
+		if err != nil {
+			return fmt.Errorf("Failed to parse image information: %v", err)
+		}
+
+		// Parse volume size into bytes, so that we can compare them.
+		volBytes, err := units.ParseByteSizeString(vol.ConfigSize())
+		if err != nil {
+			return fmt.Errorf("Failed parsing instance volume size")
+		}
+
+		// Ensure image conversion will fit into the instance volume.
+		if volBytes < imgInfo.Bytes {
+			// Convert to IEC format for nicer error.
+			imgSize := units.GetByteSizeStringIEC(imgInfo.Bytes, 2)
+			volSize := units.GetByteSizeStringIEC(volBytes, 2)
+			return fmt.Errorf("Volume size (%s) is lower then raw image size (%s)", volSize, imgSize)
+		}
+
+		// Ensure conversion supports the uploaded image format.
+		supportedImageFormats := []string{"raw", "qcow2", "vmdk"}
+		if !shared.ValueInSlice(imgInfo.Format, supportedImageFormats) {
+			return fmt.Errorf("Unsupported image format %q, allowed formats are [%s]", imgInfo.Format, strings.Join(supportedImageFormats, ", "))
+		}
+
+		// Convert uploaded image from backups directory into RAW format on the instance volume.
+		if slices.Contains(args.ConversionOptions, "format") {
+			cmd = []string{
+				"nice", "-n19", // Run with low priority to reduce CPU impact on other processes.
+				"qemu-img", "convert", "-f", imgInfo.Format, "-O", "raw", imgUploadPath, diskPath,
```

Comment:
Have a look at the use of `prlimit` in `ImageUnpack`, as I suspect it would be useful to use the same thing here.

Reply:
Thanks. I've added prlimit for both `qemu-img info` and `qemu-img convert`. 

I've limited `qemu-img convert` to 120 seconds of CPU time. While this should be more then enough for most images, I'm not sure how much time is expected to take if an image is very large (I'm having Windows images in mind, but could also be a very large root disk containing any Linux distro).